sycl/plugins/unified_runtime/ur/adapters/cuda/image.cpp

```diff
@@ -0,0 +1,1029 @@
+
+    // flags takes the normalized coordinates setting -- unnormalized is default
+    ImageTexDesc.flags = (hSampler->Props & 1) ? CU_TRSF_NORMALIZED_COORDINATES
+                                               : ImageTexDesc.flags;
+
+    // CUDA default promotes 8-bit and 16-bit integers to float between [0,1]
+    // This flag prevents this behaviour.
+    ImageTexDesc.flags |= CU_TRSF_READ_AS_INTEGER;
+
+    CUtexObject Texture;
+    RetErr = UR_CHECK_ERROR(
+        cuTexObjectCreate(&Texture, &ResourceDesc, &ImageTexDesc, nullptr));
+    *phRetImage = (ur_exp_image_handle_t)Texture;
+  } catch (ur_result_t Err) {
+    return Err;
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+  return RetErr;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urUSMPitchedAllocExp(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice,
+    const ur_usm_desc_t *pUSMDesc, ur_usm_pool_handle_t pool,
+    size_t widthInBytes, size_t height, size_t elementSizeBytes, void **ppMem,
+    size_t *pResultPitch) {
+  UR_ASSERT(hContext, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(ppMem, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pResultPitch, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT((hContext->getDevice()->get() == hDevice->get()),
+            UR_RESULT_ERROR_INVALID_CONTEXT);
+  std::ignore = pUSMDesc;
+  std::ignore = pool;
+
+  UR_ASSERT((widthInBytes > 0), UR_RESULT_ERROR_INVALID_VALUE);
+  UR_ASSERT((height > 0), UR_RESULT_ERROR_INVALID_VALUE);
+  UR_ASSERT((elementSizeBytes > 0), UR_RESULT_ERROR_INVALID_VALUE);
+
+  // elementSizeBytes can only take on values of 4, 8, or 16.
+  // small data types need to be minimised to 4.
+  if (elementSizeBytes < 4) {
+    elementSizeBytes = 4;
+  }
+  UR_ASSERT((elementSizeBytes == 4 || elementSizeBytes == 8 ||
+             elementSizeBytes == 16),
+            UR_RESULT_ERROR_INVALID_VALUE);
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  try {
+    ScopedContext Active(hDevice->getContext());
+    Result =
+        UR_CHECK_ERROR(cuMemAllocPitch((CUdeviceptr *)ppMem, pResultPitch,
+                                       widthInBytes, height, elementSizeBytes));
+  } catch (ur_result_t error) {
+    Result = error;
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  return Result;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urBindlessImagesUnsampledImageHandleDestroyExp(ur_context_handle_t hContext,
+                                               ur_device_handle_t hDevice,
+                                               ur_exp_image_handle_t hImage) {
+  UR_ASSERT(hContext, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT((hContext->getDevice()->get() == hDevice->get()),
+            UR_RESULT_ERROR_INVALID_CONTEXT);
+  UR_ASSERT(hImage, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+  RetErr = UR_CHECK_ERROR(cuSurfObjectDestroy((CUsurfObject)hImage));
+  return RetErr;
```

Comment:
Is there a benefit to having the explicit return variable, instead of:
```
return UR_CHECK_ERROR(cuSurfObjectDestroy((CUsurfObject)hImage));
```
Here and in a bunch of places below.

Reply:
Also, as `UR_CHECK_ERROR` either throws or aborts, it might be even cleaner to just return success, unless we get a compiler warning from unused return val?

Reply:
This is a good point. I've removed `RetErr` from the file and default to returning success at the end of a function, or returning error earlier and relying on `UR_CHECK_ERROR` to throw if needed.